### PR TITLE
EWPP-479: Composer 2.x compatibility.

### PR DESCRIPTION
**Pull request by yenyasinn**

### CODE REVIEW

File: src/Plugin.php [hunk 1]
```diff
@@ -58,12 +59,41 @@ public function activate(Composer $composer, IOInterface $io)
...
+            return [
+                PackageEvents::PRE_PACKAGE_INSTALL => 'prePackageInstall',
+                PackageEvents::PRE_PACKAGE_UPDATE => 'prePackageUpdate',
+            ];
+        }
+        // Events for Composer 2.
         return [
-            PackageEvents::PRE_PACKAGE_INSTALL => 'prePackageInstall',
-            PackageEvents::PRE_PACKAGE_UPDATE => 'prePackageUpdate',
+            PluginEvents::PRE_POOL_CREATE => 'prePoolCreate',
```
Comment (yenyasinn) [inline src/Plugin.php:71]: See https://github.com/composer/composer/blob/master/UPGRADE-2.0.md#detailed-differences-in-event-flow-during-dependency-resolution-composer-updates-and-installs that describes event flow in Composer 2.

File: tests/fixtures/composerProvider.yml [hunk 2]
```diff
@@ -7,7 +7,7 @@
       },
       "repositories": [
         {"packagist.org": false},
-        { "type": "path", "url": "../repositories/update/*" },
+        { "type": "path", "url": "../repositories/update/*", "canonical": false },
```
Comment (yenyasinn) [inline tests/fixtures/composerProvider.yml:10]: Composer 2 stops search repositories on the ../repositories/update/ folder. See "Repository Priorities" on the https://php.watch/articles/composer-2

File: tests/PluginComposer1Test.php [hunk 3]
```diff
@@ -59,6 +47,10 @@ public function testActivate($input, $output)
      */
     public function testPrePackageInstall($input, $output)
     {
+        if (!version_compare(PluginInterface::PLUGIN_API_VERSION, '2.0', 'lt')) {
```
Comment (yenyasinn) [inline tests/PluginComposer1Test.php:50]: Remove it, it is set in setUp

File: tests/PluginComposer1Test.php [hunk 4]
```diff
@@ -83,6 +77,10 @@ public function testPrePackageInstall($input, $output)
      */
     public function testPrePackageUpdate($input, $output)
     {
+        if (!version_compare(PluginInterface::PLUGIN_API_VERSION, '2.0', 'lt')) {
```
Comment (yenyasinn) [inline tests/PluginComposer1Test.php:80]: Remove it, it is set in setUp

File: tests/PluginComposer2Test.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        $plugin = new Plugin();
+        $plugin->activate($composer, $io);
+
+        return [
+            'event' => new PrePoolCreateEvent(
+                PluginEvents::PRE_POOL_CREATE,
+                [],
+                new Request(),
+                ['dev'],
```
Comment (yenyasinn) [inline tests/PluginComposer2Test.php:81]: Try simple `[], [],`

**Review by sergepavle — DISMISSED** [review]

File: tests/fixtures/composerProvider.yml (hunk 2 above)
Comment (brummbar) [reply tests/fixtures/composerProvider.yml:10]: Do we need this though? Which test fails if this is left unchanged?

**Review by brummbar — CHANGES_REQUESTED** [review]

File: tests/fixtures/composerProvider.yml (hunk 2 above)
Comment (yenyasinn) [reply tests/fixtures/composerProvider.yml]:
```
./vendor/bin/phpunit --testsuite composer2

There was 1 failure:

1) OpenEuropa\ComposerArtifacts\Tests\PluginIntegrationTest::testComposerCommands with data set #0 ('{\n  "name": "openeuropa/main...}\n}\n', 'install', array(array('openeuropa/dependency-1 1.0.0', 'openeuropa/dependency-2 1.0.0'), array('/main/vendor/openeuropa/depen....0.txt', '/main/vendor/openeuropa/depen...r.json', '/main/vendor/openeuropa/depen...r.json'), array('/main/vendor/openeuropa/depen....0.txt')))
Failed asserting that file "/test/composer-artifacts/tests/fixtures/main/vendor/openeuropa/dependency-1/artifact-version-1.0.0.txt" exists.

/test/composer-artifacts/tests/PluginIntegrationTest.php:84
```
Looks like composer 2 finds openeuropa/dependency-1 : 1.1.0 in repositories/update and stops to search it in another repositories (See "Repository Priorities" on the https://php.watch/articles/composer-2). But this version doesn't fit our needs "openeuropa/dependency-1": "1.0.0".

**Review by brummbar — APPROVED** [review]

**Review by sergepavle — APPROVED** [review]

**Pull request merged**